File: src/main/java/org/capturecoop/cccolorutils/chooser/gui/CCHSBHueBar.java
package org.capturecoop.cccolorutils.chooser.gui;

import org.capturecoop.cccolorutils.CCColor;
import org.capturecoop.cccolorutils.CCColorUtils;
import org.capturecoop.cccolorutils.CCHSB;
import org.capturecoop.ccutils.math.CCVector2Float;

import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.MouseMotionAdapter;
import java.awt.image.BufferedImage;
import java.util.ArrayList;

public class CCHSBHueBar extends JPanel {
    private float hue;
    private final CCColorUtils.DIRECTION direction;

    private final static int MARGIN = 10;
    private final static int SEL_MARGIN = 4;
    private final static int SEL_MARGIN_OFF = 2;
    private boolean isDragging = false;

    private BufferedImage buffer;
    private final ArrayList<ChangeListener> changeListeners = new ArrayList<>();

    public CCHSBHueBar(Color color, CCColorUtils.DIRECTION direction, boolean alwaysGrab) {
        hue = new CCHSB(color).getHue();
        this.direction = direction;
        addMouseListener(new MouseAdapter() {
            @Override
            public void mousePressed(MouseEvent mouseEvent) {
                Rectangle rect = getSelectRect();
                if(rect == null)
                    return;

                if(rect.contains(mouseEvent.getPoint()) && !alwaysGrab)
                    isDragging = true;
            }

            @Override
            public void mouseReleased(MouseEvent mouseEvent) {
                isDragging = false;
                if(alwaysGrab)
                    execute(mouseEvent.getX(), mouseEvent.getY());
            }
        });

        addMouseMotionListener(new MouseMotionAdapter() {
            @Override
            public void mouseDragged(MouseEvent mouseEvent) {
                if(isDragging || alwaysGrab)
                    execute(mouseEvent.getX(), mouseEvent.getY());
            }
        });
    }

    private void execute(int x, int y) {
        int pos = y;
        int size = getHeight();
        if(direction == CCColorUtils.DIRECTION.HORIZONTAL) {
            pos = x;
            size = getWidth();
        }
        float percentage = (pos * 100F) / size;
        hue = new CCVector2Float(percentage / 100F, 0).limitX(0.0001F, 0.9999F).getX();
        repaint();

        for(ChangeListener listener : changeListeners)
            listener.stateChanged(new ChangeEvent(this));
    }

    private int getSizeX() {
        return getWidth() - MARGIN;
    }

    private int getSizeY() {
        return getHeight() - MARGIN;
    }

    private Rectangle getSelectRect() {
        switch(direction) {
            case VERTICAL:
                int yPos = (int) (getSizeY() / (1 / hue)) + MARGIN / 2;
                return new Rectangle(SEL_MARGIN_OFF, yPos - SEL_MARGIN, getWidth() - SEL_MARGIN_OFF * 2, SEL_MARGIN * 2);
            case HORIZONTAL:
                int xPos = (int) (getSizeX() / (1 / hue)) + MARGIN / 2;
                return new Rectangle(xPos - SEL_MARGIN, SEL_MARGIN_OFF, SEL_MARGIN * 2, getHeight() - SEL_MARGIN_OFF * 2);
        }
        return null;
    }

    @Override
    public void paint(Graphics g) {
        int sizeX = getSizeX();
        int sizeY = getSizeY();

        if(buffer == null || !(buffer.getWidth() == getWidth() && buffer.getHeight() == getHeight())) {
            buffer = new BufferedImage(getWidth(), getHeight(), BufferedImage.TYPE_INT_ARGB);
            Graphics bufferGraphics = buffer.getGraphics();
            bufferGraphics.drawImage(CCColorUtils.createHSVHueBar(sizeX, sizeY, direction), MARGIN / 2, MARGIN / 2, sizeX, sizeY, this);
            bufferGraphics.dispose();
        }

        g.setColor(getBackground());
        g.fillRect(0, 0, getWidth(), getHeight());
        g.drawImage(buffer, 0, 0, this);
        g.setColor(Color.BLACK);
        g.drawRect(MARGIN / 2 - 1, MARGIN / 2 - 1, sizeX + 1, sizeY + 1);
        g.setColor(Color.GRAY);
        Rectangle rect = getSelectRect();
        g.fillRect(rect.x, rect.y, rect.width, rect.height);
    }

    public void addChangeListener(ChangeListener listener) {
        //These only fire on changes made by the user
        changeListeners.add(listener);
    }

    public float getHue() {
        return hue;
    }

    public void setHue(float hue) {
        this.hue = hue;
        repaint();
    }
}

File: src/main/kotlin/org/capturecoop/cccolorutils/chooser/gui/CCHSBHueBar.kt
package org.capturecoop.cccolorutils.chooser.gui

import org.capturecoop.cccolorutils.CCColorUtils.DIRECTION
import org.capturecoop.cccolorutils.CCColorUtils.createHSVHueBar
import org.capturecoop.cccolorutils.CCHSB
import org.capturecoop.cccolorutils.fill
import org.capturecoop.ccutils.math.CCVector2Float
import java.awt.Color
import java.awt.Graphics
import java.awt.Rectangle
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.event.MouseMotionAdapter
import java.awt.image.BufferedImage
import javax.swing.JPanel
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener

class CCHSBHueBar(color: Color?, direction: DIRECTION, alwaysGrab: Boolean) : JPanel() {
    var hue: Float = 0.0f
        set(value) {
            field = value
            repaint()
        }
    private val direction: DIRECTION
    private var isDragging = false
    private var buffer: BufferedImage? = null
    private val changeListeners = ArrayList<ChangeListener>()

    init {
        hue = CCHSB(color!!).hue
        this.direction = direction
        addMouseListener(object : MouseAdapter() {
            override fun mousePressed(mouseEvent: MouseEvent) {
                if (selectRect.contains(mouseEvent.point) && !alwaysGrab) isDragging = true
            }

            override fun mouseReleased(mouseEvent: MouseEvent) {
                isDragging = false
                if (alwaysGrab) execute(mouseEvent.x, mouseEvent.y)
            }
        })
        addMouseMotionListener(object : MouseMotionAdapter() {
            override fun mouseDragged(mouseEvent: MouseEvent) {
                if (isDragging || alwaysGrab) execute(mouseEvent.x, mouseEvent.y)
            }
        })
    }

    private fun execute(x: Int, y: Int) {
        var pos = y
        var size = height
        if (direction === DIRECTION.HORIZONTAL) {
            pos = x
            size = width
        }
        val percentage = pos * 100f / size
        hue = CCVector2Float(percentage / 100f, 0.0f).limitX(0.0001f, 0.9999f).x
        repaint()
        for (listener in changeListeners) listener.stateChanged(ChangeEvent(this))
    }

    private val sizeX: Int
        get() = width - MARGIN
    private val sizeY: Int
        get() = height - MARGIN
    private val selectRect: Rectangle
        get() = when (direction) {
            DIRECTION.VERTICAL -> {
                val yPos = (sizeY / (1 / hue)).toInt() + MARGIN / 2
                Rectangle(SEL_MARGIN_OFF, yPos - SEL_MARGIN, width - SEL_MARGIN_OFF * 2, SEL_MARGIN * 2)
            }
            DIRECTION.HORIZONTAL -> {
                val xPos = (sizeX / (1 / hue)).toInt() + MARGIN / 2
                Rectangle(xPos - SEL_MARGIN, SEL_MARGIN_OFF, SEL_MARGIN * 2, height - SEL_MARGIN_OFF * 2)
            }
        }

    override fun paint(g: Graphics) {
        val sizeX = sizeX
        val sizeY = sizeY
        if (buffer == null || !(buffer!!.width == width && buffer!!.height == height)) {
            buffer = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)
            val bufferGraphics = buffer!!.graphics
            bufferGraphics.drawImage(createHSVHueBar(sizeX, sizeY, direction), MARGIN / 2, MARGIN / 2, sizeX, sizeY, this)
            bufferGraphics.dispose()
        }
        g.color = background
        g.fillRect(0, 0, width, height)
        g.drawImage(buffer, 0, 0, this)
        g.color = Color.BLACK
        g.drawRect(MARGIN / 2 - 1, MARGIN / 2 - 1, sizeX + 1, sizeY + 1)
        g.color = Color.GRAY
        selectRect.fill(g)
    }

    fun addChangeListener(listener: ChangeListener) {
        //These only fire on changes made by the user
        changeListeners.add(listener)
    }

    companion object {
        private const val MARGIN = 10
        private const val SEL_MARGIN = 4
        private const val SEL_MARGIN_OFF = 2
    }
}